File: solutions/python3/medium/binary-search-tree-validating-the-book-collection/solution-approach1.py
# ──────────────────────────────────────────────────
# Link        https://www.hackerrank.com/contests/ti27110-sri-sai-ram/challenges/binary-search-tree-validating-the-book-collection/problem?isFullScreen=true
# Problem     Binary Search Tree - Validating the Book Collection
# Difficulty  Medium
# Subdomain   N/A
# Platform    HackerRank
# Language    python3
# Status      Accepted
# Submitted   2026-08-03, 10:36 p.m.
# ──────────────────────────────────────────────────

from collections import deque
import sys


class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_binary_tree(values):
    if not values:
        return None

    root = TreeNode(values[0])
    queue = deque([root])
    i = 1

    while queue and i < len(values):
        curr = queue.popleft()

        if i < len(values):
            curr.left = TreeNode(values[i])
            queue.append(curr.left)
            i += 1

        if i < len(values):
            curr.right = TreeNode(values[i])
            queue.append(curr.right)
            i += 1

    return root


def is_valid_bst(root, min_val=float("-inf"), max_val=float("inf")):
    if not root:
        return True

    if not (min_val < root.val < max_val):
        return False

    return is_valid_bst(root.left, min_val, root.val) and is_valid_bst(
        root.right, root.val, max_val
    )


def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return

    n = int(input_data[0])
    values = [int(x) for x in input_data[1 : n + 1]]

    root = build_binary_tree(values)

    if is_valid_bst(root):
        print("The book collection is a valid binary search tree.")
    else:
        print("The book collection is NOT a valid binary search tree.")


if __name__ == "__main__":
    main()

File: solutions/java15/medium/binary-search-tree-efficient-book-search-in-a-digital-library/solution-approach1.java
// ──────────────────────────────────────────────────
// Link        https://www.hackerrank.com/contests/ti27110-sri-sai-ram/challenges/binary-search-tree-efficient-book-search-in-a-digital-library/problem?isFullScreen=true
// Problem     Binary Search Tree - Efficient Book Search in a Digital Library
// Difficulty  Medium
// Subdomain   N/A
// Platform    HackerRank
// Language    java15
// Status      Accepted
// Submitted   2026-08-03, 10:39 p.m.
// ──────────────────────────────────────────────────

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
    }
}

public class Solution {

    private static TreeNode insertIntoBST(TreeNode root, int val) {
        if (root == null) {
            return new TreeNode(val);
        }
        if (val < root.val) {
            root.left = insertIntoBST(root.left, val);
        } else if (val > root.val) {
            root.right = insertIntoBST(root.right, val);
        }
        return root;
    }

    private static TreeNode searchBST(TreeNode root, int val) {
        TreeNode curr = root;
        while (curr != null) {
            if (curr.val == val) {
                return curr;
            } else if (val < curr.val) {
                curr = curr.left;
            } else {
                curr = curr.right;
            }
        }
        return null;
    }

    private static void inorderTraversal(TreeNode root, List<Integer> result) {
        if (root != null) {
            inorderTraversal(root.left, result);
            result.add(root.val);
            inorderTraversal(root.right, result);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextInt()) {
            return;
        }

        int n = scanner.nextInt();
        TreeNode root = null;

        for (int i = 0; i < n; i++) {
            int bookId = scanner.nextInt();
            root = insertIntoBST(root, bookId);
        }

        if (!scanner.hasNextInt()) {
            return;
        }
        int targetId = scanner.nextInt();

        TreeNode targetNode = searchBST(root, targetId);

        if (targetNode != null) {
            List<Integer> subtreeVals = new ArrayList<>();
            inorderTraversal(targetNode, subtreeVals);

            System.out.println("Book found! Subtree rooted at " + targetId + ":");
            for (int i = 0; i < subtreeVals.size(); i++) {
                System.out.print(subtreeVals.get(i) + (i == subtreeVals.size() - 1 ? "" : " "));
            }
            System.out.println();
        } else {
            System.out.print("Book not found");
        }

        scanner.close();
    }
}
